Packages/com.unity.inputsystem/InputSystem/Editor/ControlPicker/InputControlPathEditor.cs

```diff
@@ -0,0 +1,277 @@
+                    // bind to. Ignore mouse position and delta.
+                    // NOTE: We go for all types of pointers here, not just mice.
+                    .WithControlsExcluding("<Pointer>/position")
+                    .WithControlsExcluding("<Pointer>/delta")
+                    .OnCancel(
+                        operation =>
+                        {
+                            ////REVIEW: Is there a better way to do this? All we want is for the *current* UI to repaint but Unity's
+                            ////        editor API seems to have no way to retrieve the current EditorWindow from inside an OnGUI callback.
+                            ////        So we'd have to pass the EditorWindow in here all the way from the EditorWindow.OnGUI() callback
+                            ////        itself.
+                            InternalEditorUtility.RepaintAllViews();
+
+                            if (m_NeedToClearProgressBar)
+                                EditorUtility.ClearProgressBar();
+                        })
+                    .OnComplete(
+                        operation =>
+                        {
+                            if (m_NeedToClearProgressBar)
+                                EditorUtility.ClearProgressBar();
+                        })
+                    .OnApplyBinding(
+                        (operation, newPath) =>
+                        {
+                            pathProperty.stringValue = newPath;
+                            pathProperty.serializedObject.ApplyModifiedProperties();
+                            onModified();
+                        });
+
+                // If we have control paths to match, pass them on.
+                m_RebindingOperation.WithoutControlsHavingToMatchPath();
+                if (m_ControlPathsToMatch.LengthSafe() > 0)
+                    m_ControlPathsToMatch.Select(x => m_RebindingOperation.WithControlsHavingToMatchPath(x));
+
+                m_RebindingOperation.Start();
+            }
+            else if (!toggleRebind && m_RebindingOperation != null && m_RebindingOperation.started)
+            {
+                m_RebindingOperation.Cancel();
+            }
+        }
+
+        private void ShowDropdown(Rect rect)
+        {
+            if (m_PickerDropdown == null)
+            {
+                m_PickerDropdown = new InputControlPickerDropdown(
+                    m_PickerState.advancedDropdownState,
+                    path =>
+                    {
+                        pathProperty.stringValue = path;
+                        m_PickerState.manualPathEditMode = false;
+                        onModified();
+                    });
+            }
+
+            m_PickerDropdown.SetControlPathsToMatch(m_ControlPathsToMatch);
+            m_PickerDropdown.SetExpectedControlLayout(m_ExpectedControlLayout);
+
+            m_PickerDropdown.Show(rect);
+        }
+
+        private void DrawInteractivePickingProgressBar()
```

Comment:
I found that the progress bar is a vastly superior solution to the overlay I programmed before.

For one, it works everywhere regardless of where we have a pick button. So even when actions are embedded in components and shown inline in the inspector (where we have no access to the EditorWindow), we can give proper visual feedback.

Also, it make sense to have interactive picks time out, I think, and progress bars give very nice visual feedback about that.

The downside is that this change makes it harder to swallow keyboard input in the UI. We have to find a new solution for that.

Reply:
I hate to disagree, but having tried it I don't think that a progress bar is an improvement here at all. I think a progress bar is the wrong UX here, and what we had before was better. Here's why:

1. A progress bar is a convention for "The computer is doing something, wait until it is over". If you don't _know_ what this is about (for instance because you just wanted to try what that little bullseye button does), and you see a progress bar, all your trained user instincts will tell you "Ok, here's a progress bar, so whatever it's doing, i'll sit here and wait for it". Followed by "Ok, so that button did something, taking some time. I wonder what it was?".

2. I don't understand what we need the timeout for here at all?

3. The progress bar breaks the contextual association of the binding with the window you are interacting with - the old UI made it clear that you are interacting with content in the actions window, the progress bar is detached from the window and thus breaks that association.

Reply:
Let's bounce it to @shawnmcclelland. This is his home turf.

Think you have a point about the role of progress bars and I'm probably not the best to judge how isn't coming in with clear expectations. Personally, I find it much nicer than the previous UI which always seemed a bit of a kludge to me and looked even more a kludge when adapted to other places where control picking was used.

The timeout is born more out of a hate for modal things that you can't get out of except through one specific UI element. Have found myself stuck in UIs that way plenty times.

Reply:
Modal UIs are never great, and should be avoided if there is a better solution. But I'm not convinced that dismissing the modal UI after an arbitrary timeout is that better solution.

But maybe, we would not need a modal UI at all for this. I think ultimately, what this is, is a glorified input field. So what if we represent it as such? Just put in a text field, which instead of handling normal text input, process all device input as long as it is selected - and dismisses itself as soon as you click somewhere else? I'm not sure that this would work out well, but it is worth a try IMO.

Reply:
Could be worth a try. @shawnmcclelland What do you think?